### PR TITLE
feat: basic re-factor for recursion circuit, Poseidon2-wide table

### PR DESCRIPTION
**Pull request by erabinov**

### CODE REVIEW

File: recursion/core-v2/src/poseidon2_wide/air.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+//! enforce the following properties:
+//!
+
+use std::borrow::Borrow;
+
+use p3_air::{Air, BaseAir, PairBuilder};
+use p3_matrix::Matrix;
+
+use crate::builder::SP1RecursionAirBuilder;
+
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/air.rs:11]: can u group the imports that are use crate::

File: recursion/core-v2/src/lib.rs [hunk 2]
```diff
@@ -83,3 +85,20 @@ pub enum MemAccessKind {
     Read,
     Write,
 }
+
+// -------------------------------------------------------------------------------------------------
```
Comment (jtguibas) [inline recursion/core-v2/src/lib.rs:89]: nit: weird line

File: recursion/core-v2/src/poseidon2_wide/columns/mod.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+use sp1_core::utils::indices_arr;
+use sp1_derive::AlignedBorrow;
+
+use super::{NUM_INTERNAL_ROUNDS, WIDTH};
+
+pub mod permutation;
+pub mod preprocessed;
+
+/// Trait for getter methods for Poseidon2 columns.
+pub trait Poseidon2<'a, T: Copy + 'a> {
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/columns/mod.rs:13]: why do we need this trait

File: recursion/core-v2/src/poseidon2_wide/columns/mod.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    fn state_var(&self) -> &[T; WIDTH];
+    fn internal_rounds_s0(&self) -> &[T; NUM_INTERNAL_ROUNDS - 1];
+    fn s_box_state(&self) -> Option<&[T; WIDTH]>;
+}
+
+/// Trait for setter methods for Poseidon2 columns. Only need the memory columns are populated mutably.
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/columns/mod.rs:19]: same here

File: recursion/core-v2/src/poseidon2_wide/columns/mod.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        Option<&mut [T; WIDTH]>,
+    ) {
+        match self {
+            Poseidon2MutEnum::P2Degree3(p) => p.get_cols_mut(),
+            Poseidon2MutEnum::P2Degree9(p) => p.get_cols_mut(),
+        }
+    }
+}
+
+pub const NUM_POSEIDON2_DEGREE3_COLS: usize = size_of::<Poseidon2Degree3<u8>>();
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/columns/mod.rs:82]: const's should almost always be at the top of the file

Comment (jtguibas) [reply recursion/core-v2/src/poseidon2_wide/columns/mod.rs:82]: i know copied but

File: recursion/core-v2/src/poseidon2_wide/columns/mod.rs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    }
+}
+pub const POSEIDON2_DEGREE3_COL_MAP: Poseidon2Degree3<usize> = make_col_map_degree3();
+
+/// Struct for the poseidon2 chip that contains sbox columns.
+#[derive(AlignedBorrow, Clone, Copy)]
+#[repr(C)]
+pub struct Poseidon2Degree3<T: Copy> {
+    pub permutation_cols: PermutationSBox<T>,
+}
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/columns/mod.rs:97]: lack of space weird

File: recursion/core-v2/src/poseidon2_wide/columns/permutation.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    pub internal_rounds_s0: [T; NUM_INTERNAL_ROUNDS - 1],
+}
+
+#[derive(AlignedBorrow, Clone, Copy)]
+#[repr(C)]
+pub struct PermutationSBoxState<T: Copy> {
+    pub sbox_state: [T; max(WIDTH, NUM_INTERNAL_ROUNDS)],
+}
+
+/// Trait that describes getter functions for the permutation columns.
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/columns/permutation.rs:30]: unclear why these traits are unecessary to me

File: recursion/core-v2/src/poseidon2_wide/columns/preprocessed.rs [hunk 8]
```diff
@@ -0,0 +1,28 @@
+use sp1_derive::AlignedBorrow;
+
+use crate::mem::MemoryPreprocessedColsNoVal;
+use crate::poseidon2_wide::WIDTH;
+
+// pub const POSEIDON2_MEMORY_PREPROCESSED_WIDTH: usize = size_of::<MemoryPreprocessedkkj<u8>>();
```
Comment (jtguibas) [inline recursion/core-v2/src/poseidon2_wide/columns/preprocessed.rs:6]: nit: fix

File: recursion/core-v2/src/lib.rs (hunk 2 above)
Comment (erabinov) [reply recursion/core-v2/src/lib.rs:89]: is it the empty line or the `// ------` comment? The comment is just following Tej's formatting in the rest of the file.

File: recursion/core-v2/src/runtime/mod.rs [hunk 9]
```diff
@@ -89,6 +88,18 @@ pub struct Runtime<F: PrimeField32, EF: ExtensionField<F>, Diffusion> {
...
+        Poseidon2<
+            F,
+            Poseidon2ExternalMatrixGeneral,
+            Diffusion,
+            PERMUTATION_WIDTH,
+            POSEIDON2_SBOX_DEGREE,
+        >,
+    >,
+
+    // p2_hash_state: [F; PERMUTATION_WIDTH],
```
Comment (jtguibas) [inline recursion/core-v2/src/runtime/mod.rs:102]: remove

**Review by jtguibas — APPROVED** [review]

**Pull request merged**